source/geod24/concurrency.d

```diff
@@ -1012,56 +1178,154 @@ private:
 
         Start the scheduling loop
 
-        Param:
-            pos = m_fibers index of the Fiber to execute first
-
     ***********************************************************************/
 
-    void dispatch(size_t pos = 0)
+    void dispatch ()
     {
-        import std.algorithm.mutation : remove;
+        thisScheduler(this);
+        scope (exit) thisScheduler(null);
+
+        MonoTime earliest_timeout;
 
-        assert(pos < m_fibers.length);
-        while (m_fibers.length > 0)
+        while (!this.readyq.empty() || !this.wait_list.empty())
         {
-            // Is Fiber waiting on a FiberBinarySemaphore?
-            if (auto sem = m_fibers[pos].sem)
+            while (!readyq.empty())
```

Comment:
I'm a little bit confused about the nested loop. Can't you just remove it? The outer while loop handles everything.

Reply:
We would need to replace the inner `while` with `if` and add an additional check to the `if` below to not avoid sleeping while ready queue is not empty. I guess this could work

```diff
diff --git a/source/geod24/concurrency.d b/source/geod24/concurrency.d
index 947835d..909ef15 100644
--- a/source/geod24/concurrency.d
+++ b/source/geod24/concurrency.d
@@ -1190,7 +1190,7 @@ private:
 
         while (!this.readyq.empty() || !this.wait_list.empty())
         {
-            while (!readyq.empty())
+            if (!readyq.empty())
             {
                 InfoFiber cur_fiber = this.readyq.front();
                 this.readyq.removeFront();
@@ -1233,8 +1233,7 @@ private:
                 if (MonoTime.currTime >= earliest_timeout || this.sem.tryWait())
                     earliest_timeout = wakeFibers();
diff --git a/source/geod24/concurrency.d b/source/geod24/concurrency.d
index 947835d..909ef15 100644
--- a/source/geod24/concurrency.d
+++ b/source/geod24/concurrency.d
@@ -1190,7 +1190,7 @@ private:
 
         while (!this.readyq.empty() || !this.wait_list.empty())
         {
-            while (!readyq.empty())
+            if (!readyq.empty())
             {
                 InfoFiber cur_fiber = this.readyq.front();
                 this.readyq.removeFront();
@@ -1233,8 +1233,7 @@ private:
                 if (MonoTime.currTime >= earliest_timeout || this.sem.tryWait())
                     earliest_timeout = wakeFibers();
             }
-
-            if (!this.wait_list.empty())
+            else if (!this.wait_list.empty())
             {
                 Duration time_to_timeout = earliest_timeout - MonoTime.currTime;
 
@@ -1243,10 +1242,10 @@ private:
                     this.sem.wait();
                 else if (time_to_timeout > 0.seconds)
                     this.sem.wait(time_to_timeout);
-            }
 
-            // OS Thread woke up populate ready queue
-            earliest_timeout = wakeFibers();
+                // OS Thread woke up populate ready queue
+                earliest_timeout = wakeFibers();
+            }
         }
```